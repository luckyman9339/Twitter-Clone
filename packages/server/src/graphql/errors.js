import { ApolloError } from 'apollo-server-express';

export class UnknownError extends ApolloError {
  constructor(properties) {
    super('An unknown error has occurred', 'UNKNOWN', properties);

    Object.defineProperty(this, 'name', { value: 'UnknownError' });
  }
}

export class AlreadyInUseError extends ApolloError {
  constructor(message, field) {
    super(message, 'ALREADY_IN_USE', { field });

    Object.defineProperty(this, 'name', { value: 'AlreadyInUseError' });
  }
}
