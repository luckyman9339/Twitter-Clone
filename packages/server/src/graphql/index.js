import fs from 'fs';
import path from 'path';
import { makeExecutableSchema } from 'graphql-tools';
import { fileLoader, mergeTypes, mergeResolvers } from 'merge-graphql-schemas';

/*
 * Create Types
 */

const typesArray = fileLoader(path.join(__dirname, '.'), {
  recursive: true,
  extensions: ['.graphql'],
});

const typeDefs = mergeTypes(typesArray);

/*
 * Create Resolvers
 */

// Seems like we can't import the resolvers with a blob on backpack?
const resolversArray = [
  require('./schema/user/resolvers').default,
  require('./schema/tweet/resolvers').default,
  require('./schema/feed/resolvers').default,
];
const resolvers = mergeResolvers(resolversArray);

/*
 * Create Schema
 */

export default makeExecutableSchema({ typeDefs, resolvers });
