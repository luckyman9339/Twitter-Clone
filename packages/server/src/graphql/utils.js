export const decodeCursor = cursor => {
  const content = new Buffer(cursor, 'base64').toString('binary');
  const [after, order] = content.split(';');
  return { after: new Date(after), order };
};

export const encodeCursor = ({ after, order }) => {
  const input = [new Date(after).toISOString(), order].join(';');
  return Buffer.from(input).toString('base64');
};

export const buildEdgesForTweetsPagination = (tweets, order) =>
  tweets.map(tweet => {
    const type = tweet.retweetForTweetId
      ? 'RETWEET'
      : tweet.likeForTweetId
      ? 'LIKE'
      : 'TWEET';

    const node = { ...tweet, type };

    return {
      cursor: encodeCursor({
        after: tweet.created_at,
        order,
      }),
      node,
    };
  });
