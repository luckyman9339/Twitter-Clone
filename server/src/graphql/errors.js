import { createError } from 'apollo-errors';

// Mask any internal errors
export const UnknownError = createError('UnknownError', {
  message: 'An unknown error has occurred',
});

// User should be logged in but isn't
export const UnauthorizedError = createError('UnauthorizedError', {
  message: 'You must login to do that',
});
