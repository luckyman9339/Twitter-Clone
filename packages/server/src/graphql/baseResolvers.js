import { ApolloError, AuthenticationError } from 'apollo-server-express';
import { createResolver } from 'apollo-resolvers';
import { UnknownError } from './errors';

export const baseResolver = createResolver(null, (root, args, context, err) => {
  if (err instanceof ApolloError || process.env.NODE_ENV !== 'production') {
    return err;
  }

  return new UnknownError();
});

export const isAuthenticatedResolver = baseResolver.createResolver(
  (root, args, context) => {
    const { user } = context;
    if (!user) throw new AuthenticationError('You need to login to do this!');
  }
);
