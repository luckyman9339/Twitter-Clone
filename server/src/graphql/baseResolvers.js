import { isInstance } from 'apollo-errors';
import { createResolver } from 'apollo-resolvers';
import { UnknownError, UnauthorizedError } from './errors';

export const baseResolver = createResolver(null, (root, args, context, err) => {
  if (isInstance(err) || process.env.NODE_ENV !== 'production') {
    return err;
  }

  return new UnknownError({ data: { name: err.name } });
});

export const isAuthenticatedResolver = baseResolver.createResolver(
  (root, args, context) => {
    const { user } = context;
    if (!user) throw new UnauthorizedError();
  }
);
