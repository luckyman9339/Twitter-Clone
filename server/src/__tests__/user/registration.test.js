import { graphql } from 'graphql';
import schema from '../../graphql';
import knex from '../../db/knex';

const request = (query, { context, variables }) => {
  return graphql(
    schema,
    query,
    undefined,
    {
      ...context,
    },
    variables
  );
};

describe('registration', () => {
  beforeEach(() => {
    return knex.migrate.rollback().then(() => knex.migrate.latest());
    // .then(() => knex.seed.run());
  });

  afterEach(() => {
    return knex.migrate.rollback();
  });

  afterAll(() => {
    return knex.destroy();
  });

  it('should register an user', async () => {
    const query = `
      mutation {
        registerUser(
          input: {
            username: "rafael"
            password: "aaa"
            email: "test@email.com"
            name: "Rafael"
          }
        ) {
          username
          email
          name
        }
      }
    `;

    const result = await request(query, {});
    expect(result).toMatchSnapshot();
  });

  it('should not register an user with the same username', async () => {
    const query = `
      mutation {
        registerUser(
          input: {
            username: "rafael"
            password: "aaa"
            email: "test@email.com"
            name: "Rafael"
          }
        ) {
          username
          email
          name
        }
      }
    `;

    const result = await request(query, {});
    expect(result).toMatchSnapshot();

    const resultTwo = await request(query, {});
    expect(resultTwo).toMatchSnapshot();
  });
});
