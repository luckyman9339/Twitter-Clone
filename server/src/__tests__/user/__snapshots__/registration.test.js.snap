// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration should not register an user with the same username 1`] = `
Object {
  "data": Object {
    "registerUser": Object {
      "email": "test@email.com",
      "name": "Rafael",
      "username": "rafael",
    },
  },
}
`;

exports[`registration should not register an user with the same username 2`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: This username is already in use],
  ],
}
`;

exports[`registration should register an user 1`] = `
Object {
  "data": Object {
    "registerUser": Object {
      "email": "test@email.com",
      "name": "Rafael",
      "username": "rafael",
    },
  },
}
`;
